crates/vm/backends/levm.rs

```diff
@@ -0,0 +1,314 @@
+
+        if initial_account_state.is_none() {
+            // New account, update everything
+            let new_account = AccountUpdate {
+                address: *new_state_account_address,
+                removed: new_state_account.is_empty(),
+                info: Some(AccountInfo {
+                    code_hash: code_hash(&new_state_account.info.bytecode),
+                    balance: new_state_account.info.balance,
+                    nonce: new_state_account.info.nonce,
+                }),
+                code: Some(new_state_account.info.bytecode.clone()),
+                added_storage: new_state_account
+                    .storage
+                    .iter()
+                    .map(|(key, storage_slot)| (*key, storage_slot.current_value))
+                    .collect(),
+            };
+
+            account_updates.push(new_account);
+            continue;
+        }
+
+        // This unwrap is safe, just checked upside
+        let initial_account_state = initial_account_state.unwrap();
+        let mut account_update = AccountUpdate::new(*new_state_account_address);
+
+        // Account state after block execution.
+        let new_state_acc_info = AccountInfo {
+            code_hash: code_hash(&new_state_account.info.bytecode),
+            balance: new_state_account.info.balance,
+            nonce: new_state_account.info.nonce,
+        };
+
+        // Compare Account Info
+        if initial_account_state != new_state_acc_info {
+            account_update.info = Some(new_state_acc_info.clone());
+        }
+
+        // If code hash is different it means the code is different too.
+        if initial_account_state.code_hash != new_state_acc_info.code_hash {
+            account_update.code = Some(new_state_account.info.bytecode.clone());
+        }
+
+        let mut updated_storage = HashMap::new();
+        for (key, storage_slot) in &new_state_account.storage {
+            // original_value in storage_slot is not the original_value on the DB, be careful.
+            let original_value = current_db
+                .get_storage_at_hash(block_hash, *new_state_account_address, *key)
+                .unwrap()
+                .unwrap_or_default(); // Option inside result, I guess I have to assume it is zero.
+
+            if original_value != storage_slot.current_value {
+                updated_storage.insert(*key, storage_slot.current_value);
+            }
+        }
+        account_update.added_storage = updated_storage;
+
+        account_update.removed = new_state_account.is_empty();
+
+        if account_update != AccountUpdate::new(*new_state_account_address) {
+            account_updates.push(account_update);
+        }
+    }
+    account_updates
+}
+
+/// Calls the eip4788 beacon block root system call contract
+/// More info on https://eips.ethereum.org/EIPS/eip-4788
+pub fn beacon_root_contract_call_levm(
+    store_wrapper: Arc<StoreWrapper>,
```

Comment:
Why is this a `Arc<StoreWrapper`? Shouldn't it be a `db: Arc<dyn LevmDatabase>` like the one in `execute_tx_levm`?
Just asking, maybe it got left out.
```suggestion
    db: Arc<dyn LevmDatabase>,
```